### ISSUE TITLE
Checkboxes issues in elements editors forms in dev branch (+ possible fix)

### ISSUE THREAD

**Issue by jtkech** [OrchardCMS/Orchard#6302]
In the dev branch

1. Open an element editor that have a checkbox option.
2. Check the option (if not already done) and save.
2. Re-open the editor, uncheck the option and save.
3. Re-open the editor, the option is still checked.

In 1.9.x, there is this PR #6301 that solves a same kind of issue #5593 due to an option that is initially checked... But in the dev branch it concerns all checkboxes (defined through the forms api), whether initially checked or not...

The problem is that if a checkbox is not checked, no value is posted on submission. So, here, the beginning of a possible solution in the dev branch by modifying `FormsElementDriver.cs`.

        private void UpdateElementProperty(dynamic formElementShape, ElementEditorContext context) {
            var name = (string)formElementShape.Name;
            if (name != null) {
                var value = context.ValueProvider.GetValue(context.Prefix + name);
                if (value != null) {
                    context.Element.Data[name] = value.AttemptedValue;
                }

                // Added code...
                else if(formElementShape.Metadata.Type == "Checkbox") {
                    if (formElementShape.Value == "true")
                        context.Element.Data[name] = "false";
                    else if (formElementShape.Default != null)
                        context.Element.Data[name] = formElementShape.Default;
                }
            }
        }

Here, if nothing is posted, the mainly used "true" value of the `shape.Value` property is tested, then "false" is assigned... To implement other cases a `shape.Default` property can be simply added through the `DescribeForm()`.

Best

**Comment by sfmskywalker** [comment]
I don't think this is an issue with the Forms API, unless you can reproduce this issue with the Workflows module for example (which also uses the Forms API). Instead, I suspect this has been introduced when refactoring the way form values are assigned to element data, somewhere in the ElementController.

**Comment by jtkech** [comment]
@sfmskywalker 

You're right, that's why i first modified `ElementController.Update()` where the **element could be activated with some previous data values**, because when a checkbox is unchecked no new value is posted . Before, the `Request.Form` was used, so, with only the posted values.

I could fix it in the `Update()` action where i've used the `removeNonExistingItems` option to true with the `ElementDataDictionary Combine()` extension:

`var data = elementData.Combine(ElementDataHelper.Deserialize(sessionState.ElementEditorData), true);`

But, as in the **1.9.x branch**, this doesn't fix the issue where **you can't define through DescribeForm() a checkbox that is initially checked**.

That's why, in my previous comment, i've proposed another fix that works, so if it can help. This fix doesn't modify the `Forms API`, only the `FormsElementDriver.UpdateElementProperty()` in the `Orchard.Layouts` module.

Best

**Comment by sfmskywalker** [comment]
Ah I see. Sorry, small screen size ;) But yes, if you could fix that as proposed then that'd be awesome. Please test if all standard operations still behave as expected. I'll make a note of adding some automated tests to protect these things from regressions. Thanks man, really appreciate the time and effort you're spending on this (and other aspects) :)

**Comment by jtkech** [comment]
@sfmskywalker thanks.

Just tried with the **workflows editor** through the create user form, it works as before without the 1st issue. But, here also, we could have the **2nd issue** if we define a checkbox that is initially checked.

So:

1. With the little fix in ElementController, it's as before, but only the 1st issue is fixed, and you can't use a checkbox that is initially checked (and the related element property with a default value of true).

2. If we apply the first proposed fix, the 2 issues seems to be fixed but only for the layout elements.

3. Try to find a general solution (see below) to always post a value with a "boolean" checkbox (whose value attribute = "true"). (This give me the idea to implement a 2 values checkbox element...).

In asp source code, i've seen a checkbox helper that add an hidden field to post a "false" value. I could do it but it was not enough, maybe ok with a regular mvc pattern, but i also needed to remove the hidden field when the checkbox is checked, this not to post 2 values...

So, i've found a solution with a few lines of code in `EditorShapes.cs`, but i also needed to add a few lines of jquery code in `base.js`. Seems to work but needs more tests, i will propose a PR soon.

Best

**Comment by jtkech** [comment]
@sfmskywalker 

For infos I've proposed this PR #6308 with some jquery code. @sebastienros already asked me some questions. Not sure it's a good fix but it's the only way i've found to be more general with existing implementations (see my previous comment above).

Best

**Comment by sfmskywalker** [comment]
I remember this issue. As per @jtkech 's earlier comment, this was originally fixed by simply removing non-posted items from the element's data dictionary. However, the way this should be fixed is the same way the Checkbox HTML helper is implemented, by generating a hidden input with the same name as the checkbox field that posts "false" as its value. This way, when the checkbox is not checked, "false" will be posted. If the checkbox is checked, "false,true" will be posted (or the other way around), and "true" will be used. If that doesn't work as per @jtkech 's comments in PR #6308, then that is what we need to fix, since it should work with the modelbinder. Perhaps the order of the values matter, or perhaps it has to do with the underlying value provider. For example, `NameValueCollection` can store multiple values per entry in the form of a Values array, but a Dictionary probably just returns the posted string value ("true,false"). That should be fixable.

To summarize, I think we need to include the hidden input field by updating the Checkbox form shape (Orchard.Forms).

**Comment by jtkech** [comment]
Just re-tried #6308, it works but with a little jquery code where i add an hidden field only if needed. Notice that #6339 also solves the issue through `UpdateElementProperty()` without an hidden field, but more specifically for layout element editors (maybe sufficient?).

But maybe using some jquery code is not the best solution, and its better to see around model binding as you said. But as i remember, modules that use Orchard.Forms checkboxes don't use model binding in the same way, e.g workflows editors.

So, here we would need to solve things step by step as needed. E.g right now this is not needed for workflows because the issue is only when a checkbox is initially checked, and currently this is not the case. So, maybe #6339 would be the first step.

Best
